pkg/datapath/loader/loader.go

```diff
@@ -477,9 +477,26 @@ func attachCiliumNet(ep datapath.Endpoint, spec *ebpf.CollectionSpec) error {
 func attachNetworkDevices(cfg *datapath.LocalNodeConfiguration, ep datapath.Endpoint, spec *ebpf.CollectionSpec) error {
 	devices := cfg.DeviceNames()
 
-	// Selectively attach bpf_host to cilium_wg0.
-	if option.Config.NeedBPFHostOnWireGuardDevice() {
+	// Selectively attach/detach bpf_host to cilium_wg0.
+	// If WireGuard is not enabled, then either the interface
+	// does not exists or it has been removed.
+	switch {
+	case option.Config.NeedBPFHostOnWireGuardDevice():
+		// Program needed, attach bpf_host to cilium_wg0.
 		devices = append(devices, wgTypes.IfaceName)
+	case option.Config.EnableWireguard:
```

Comment:
I think I would move this case to the `replaceWireguardDatapath` function, then this function keeps it's purpose of attaching bpf_host when needed, and `replaceWireguardDatapath` can remove progs if no longer needed. And IIUC, all the attaching and detaching to `cilium_wg0` will move to `replaceWireguardDatapath` going forward anyway.

Reply:
Agree, but would this also apply then to https://github.com/cilium/cilium/pull/38187?
In v1.17 I need to land the logic for the removal of `cil_from_wireguard` (introduced in v1.18), and should be performed as soon as `cil_from_netdev` is attached.

Reply:
hm, yeah the `from-wireguard` removal for the downgrade scenario to 1.17 probably needs to live in `attachNetworkDevices`, and happen immediately after `cil_from_netdev` is attached. But I think that is fine, given it's a special case to handle downgrades. But IIUC the handling of the `to-wireguard` prog is completely separated from the handling of the `from-wireguard` correct? So this PR is somewhat independent of the other changes right?

Reply:
Yep, agree. Apologies for deleting my comment, I did it for this exact same reason: I realized they're separate changes and can live on their own. I noticed later you reply.

Just applied the suggested changes :pray: